src/valgrind_selftest/valgrind_selftest.cpp

```diff
@@ -140,6 +141,57 @@ void test_conditional_branch_on_poisoned_bit() {
    }
 }
 
+void test_clang_conditional_jump_on_bare_metal_ct_mask() {
+   Botan::AutoSeeded_RNG rng;
+   const auto poisoned_byte = rng.next_byte();
+   Botan::CT::poison(&poisoned_byte, sizeof(poisoned_byte));
+
+   std::array<uint8_t, 16> output_bytes;
+   std::memset(output_bytes.data(), 0x42, sizeof(output_bytes));
+
+   // This mimicks what went wrong in Kyber's secret message expansion
+   // that was found by PQShield in Kyber's reference implementation and
+   // was fixed in https://github.com/randombit/botan/pull/4107.
+   //
+   // Certain versions of Clang, namely 15, 16, 17 and 18 (maybe more) with
+   // specific optimization flags (-Os, -O1, -O2 -fno-vectorize, ...) do
+   // realize that `poisoned_mask` can only ever be all-zero or all-one and
+   // conditionally jump over the loop execution below.
+   //
+   // See: https://pqshield.com/pqshield-plugs-timing-leaks-in-kyber-ml-kem-to-improve-pqc-implementation-maturity/
+   const uint8_t poisoned_mask = -static_cast<uint8_t>(poisoned_byte & 1);
+   for(size_t i = 0; i < sizeof(output_bytes); ++i) {
+      output_bytes[i] &= poisoned_mask;
+   }
+
+   // Unpoison output_bytes to safely print them. The actual side channel
+   // happened above.
+   Botan::CT::unpoison(output_bytes.data(), output_bytes.size());
+   std::cout << Botan::hex_encode(output_bytes) << std::endl;
+}
+
+// Similar to test_clang_conditional_jump_on_bare_metal_ct_mask() but with Botan's CT::Masks.
+// Should not fail, because CT::Mask uses CT::value_barrier().
+void regression_test_conditional_jump_in_ct_mask() {
+   Botan::AutoSeeded_RNG rng;
+   const auto poisoned_byte = rng.next_byte();
+   Botan::CT::poison(&poisoned_byte, sizeof(poisoned_byte));
+
+   std::array<uint8_t, 16> output_bytes;
+   std::memset(output_bytes.data(), 0x42, sizeof(output_bytes));
+
+   // Before the introduction of CT::value_barrier, this did generate a
+   // conditional jump when compiled with clang using certain compiler
+   // optimizations. See the test case above for further details.
+   auto poisoned_mask = Botan::CT::Mask<uint8_t>::expand(poisoned_byte & 1);
+   for(size_t i = 0; i < sizeof(output_bytes); ++i) {
+      output_bytes[i] = poisoned_mask.select(output_bytes[i], 0);
+   }
```

Comment:
If you comment-out the gist of `CT::value_barrier()` this also triggers valgrind under the same conditions explained above: https://github.com/randombit/botan/pull/4182/commits/487eb5034c76eb2dd80b13f83d6a42b526da1179#r1666833485